design/reproducible-metal3-dev-env.md

```diff
@@ -0,0 +1,144 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Reproducible metal3-dev-env
+
+## Status
+
+provisional
+
+## Summary
+
+The [metal3-dev-env](https://github.com/metal3-io/metal3-dev-env) is used by
+developers working on metal3 components, for all CI builds and integration
+tests, and it is also something of a reference for users deploying metal3. 
+
+At the moment, the environment is quite far from being reproducible. In other
+words, running `make` in metal3-dev-env at the same commit on two different
+days could lead to different results and different software being run in the
+environment, regardless of any changes to metal3 projects. In several places,
+the Bash scripts in metal3-dev-env get the latest version of external software
+and automatically pull this into the environment, meaning that upstream
+releases change or break the development environment.
+
+The purpose of this proposed change is to pin external dependencies of metal3
+in the metal3-dev-env, and use automated tooling, such as
+[Renovate](https://github.com/renovatebot/renovate) to keep these dependencies
+up to date via pull requests. Then updates to external software which may break
+the environment can be gated at pull request time, when an integration test is
+run. Furthermore, if there are external changes which flakily break the build
+(and get past an integration test), it is simple enough to revert the merge
+commit to fix the environment.
+
+## Motivation
+
+On weeks commencing 11th April 2021 and 18th April 2021, there were several
+software releases and wider infrastructure issues which led to build failures
+on the master branch:
+
+* Minikube v1.19 was released and automatically pulled into our environment. It
+  appears to have a flaky issue with instantiating an IP address for the
+  minikube VM in our environment, which we did not see in v1.18.
+* A kubelet bug in Kubernetes v1.20.4 occasionally breaks builds with a
+  multi-node control plane
+* A change to tripleo-repos script was pulled in automatically by the
+  Dockerfile which builds vbmc, and broke all builds and integration tests
+* A network problem between the CI infrastructure and images.rdoproject.org led
+  to build timeouts when the Ironic Python Agent was downloaded
```

Comment:
This last one doesn't seem to be relevant, IIUC the proposal is to pin versions, and implement a way to automate bumping those versions - which doesn't solve the issue that sometimes network outages happen?

Reply:
@hardys, thanks for taking a look at the proposal. Yes, you are right. The network issue that we had drew my attention to the fact that we're downloading from https://images.rdoproject.org/centos8/master/rdo_trunk/current-tripleo/, where fresh tarballs are pushed and not versioned, but the network troubles themselves would not have been ameliorated by any of the suggestions in this proposal. I've removed this line.

(As an aside, for CI builds, I'm now downloading the IPA tarball from a mirror closer to the infrastructure so that we don't run into this problem again. Sweden -> US West was quite a way to go for the IPA! But the IPA is still not versioned.)

Reply:
Right, we have a CACHEURL variable which can point to a locally mirrored copy https://github.com/metal3-io/ironic-ipa-downloader/blob/master/get-resource.sh#L26 so perhaps doing that in CI would be a good idea, ideally with some automated way to test any new version before updating the cached copy.